src/NumericalAlgorithms/SphericalHarmonic/ComplexDataSlice.hpp

```diff
@@ -0,0 +1,171 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <complex>
+#include <cstddef>
+
+#include "DataStructures/ComplexDataVector.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+
+enum ComplexSliceKind {
+  interleaved,
+  real_slices,
+};
+
+/// \brief A storage container for storing sub-vectors ("slices") of a
+/// ComplexDataVector.
+///
+/// \details Abstraction class which deals with the memory management for
+/// possible representations of complex numbers, chosen by the
+/// `ComplexSliceKind` parameter. Two options are offered:
+/// 1. `ComplexSliceKind::interleaved`: the real and imaginary values alternate
+/// in memory. Internally, this is managed by a `ComplexDataVector`.
+/// 2. `ComplexSliceKind::real_slices`: the real and imaginary values are two
+/// separate, contiguous blocks of double in memory. Internally, this is managed
+/// by a pair of `DataVector`s, which then copy back to the pointer to the
+/// original data either when the `get_ComplexDataVector()` routine is called to
+/// access the complex data, or when it is forced by a call to
+/// `copy_back_to_source`.
+template <ComplexSliceKind Kind>
+class ComplexDataSlice {
+ private:
+  size_t size_;
+
+  static const size_t stride_ = Kind == ComplexSliceKind::real_slices ? 1 : 2;
+
+  // In either case, we want to avoid unnecessary copies, so we track whether
+  // the data has been copied from one representation to the other.
+  bool real_slices_up_to_date_;
+  DataVector real_slice_;
+  DataVector imag_slice_;
+
+  double* data_real_;
+  double* data_imag_;
+
+  bool complex_slice_up_to_date_;
+  ComplexDataVector complex_slice_;
+
+ public:
+  /// \brief The Kind determines the internal data representation
+  static const ComplexSliceKind slice_kind = Kind;
+
+  /// \brief The internal storage type
+  using sliceElementType = std::complex<double>;
+
+  /// \brief Construct a slice which starts at index `offset` of the supplied
+  /// vector, and extends for `size` elements
+  ///
+  /// \details A copy occurs in the case of a `real_slices` kind of slice.
+  ComplexDataSlice(gsl::not_null<ComplexDataVector*> vector, size_t size,
+                   size_t offset = 0) noexcept;
+
+  /// \brief Construct a slice which starts at pointer `start` and extends for
+  /// `size` elements. Need not be a part of a ComplexDataVector.
+  ///
+  /// \details A copy occurs in the case of a `real_slices` kind of slice.
+  ComplexDataSlice(std::complex<double>* start, size_t size) noexcept;
+
+  // For the lifetime of the data slice, it points to the same portion of a
+  // single vector, default move assign behavior would contradict that
```

Comment:
So should this class be copyable or movable at all? This comment makes me think "no" but it doesn't make the intent of the overall class very clear. I think this should be documented in the main dox comment clarifying what things this class guarantees and what it leaves as UB